File: quant1x/datasets/kline_minute.cpp
#include <quant1x/datasets/kline_minute.h>
#include <quant1x/datasets/kline_raw.h>
#include <ranges>

#include <quant1x/pandas/rule.h>

namespace datasets {

    namespace {

        void save_kline(const std::string &filename, const std::vector<MinuteKLine>& values) {
            util::check_filepath(filename, true);
            io::CSVWriter writer(filename);
            writer.write_row("Date", "Open", "Close", "High", "Low", "Volume", "Amount", "Up", "Down", "Datetime", "AdjustmentCount");
            for (const auto &row: values) {
                writer.write_row(row.Date, row.Open, row.Close, row.High, row.Low, row.Volume, row.Amount,
                                 row.Up, row.Down, row.Datetime, row.AdjustmentCount);
            }
        }

        void calculate_pre_adjust(std::vector<MinuteKLine> &klines, const exchange::timestamp &startDate, const std::vector<level1::XdxrInfo> &dividends) {
            if(klines.empty()) {
                return;
            }
            // 最后一根K线的日期
            auto const& last_day = klines[klines.size()-1].Date;
            // 转成时间戳且对齐时间
            auto const& ts_last_day = exchange::timestamp::parse(last_day).pre_market_time();
            // 计算最后一根K线的下一个交易日的日期, 除权除息是不包括除权除息当日的, 所以要计算下一个交易日与除权除息的列表去匹配
            // 300773拉卡拉, 2025年6月6日除权, 数据公布于6月3日之前, 那么在6月6日之前的6月4日收盘前是不能除权除息的，6月5日收盘可以除权
            auto const& last_day_next = exchange::next_trading_day(ts_last_day).only_date();
            auto start_date = startDate.only_date();
            auto xdxr_infos = dividends | std::views::filter([&last_day_next](const level1::XdxrInfo & x) {return last_day_next >= x.Date && x.Category == 1;});
            //int times = 0; // 除权除息次数
            size_t count = std::ranges::distance(xdxr_infos); // 除权除息总次数
            // 时间越早的记录除权除息次数越多, 第一条数据时时总的除权除息次数
            auto times = count;
            for(auto const & info : xdxr_infos) {
                if(info.Date <= start_date) {
                    // 除权除息数据在日线第一条数据之前, 也就是ipo上市日期之前的数据, 不能用作复权
                    //continue;
                } else {
                    auto [m, a] = info.adjustFactor();
                    auto klines_size = klines.size();
                    for (size_t i = 0; i < klines_size; ++i) {
                        auto kl = &(klines[i]);
                        if (kl->Date >= info.Date) {
                            break;
                        }
                        if (kl->Date < info.Date) {
                            kl->Open = kl->Open * m + a;
                            kl->Close = kl->Close * m + a;
                            kl->High = kl->High * m + a;
                            kl->Low = kl->Low * m + a;
                            // 成交量复权
                            // 1. 计算均价
                            auto ap = kl->Amount / kl->Volume;
                            // 2. 均价复权
                            ap = ap * m + a;
                            // 3. 以成交金额为基准, 用复权均价计算成交量
                            kl->Volume = kl->Amount / ap;
                            kl->AdjustmentCount += 1;
                        }
                    }
                }
                --times;
                (void)times;
            }
        }

        // config::MinuteKLineConfig kline_config() {
        //     config::MinuteKLineConfig config{};
        //     auto const &local_cfg = config::global_config().data.cache.kline;
        //     if (local_cfg.size() != 1) {
        //         throw std::runtime_error("kline config size must be exactly one");
        //     }
        //     const auto minute_kline_config = local_cfg.begin();
        //     const auto key = minute_kline_config->first;
        //     const auto value = minute_kline_config->second;
        //     const auto d = pandas::ParseTimeRule(key);
        //     const auto minutes = std::chrono::duration_cast<std::chrono::minutes>(d);
        //     config.minutes = minutes.count();
        //     config.frequency = key;
        //     config.enabled = value;
        //     return config;
        // }

    }

    void MinuteKLine::adjust(double m, double a, int number) {
        Open = Open * m + a;
        Close = Close * m + a;
        High = High * m + a;
        Low = Low * m + a;
        // 成交量复权
        // 1. 计算均价
        auto ap = Amount / Volume;
        // 2. 均价复权
        ap = ap * m + a;
        // 3. 以成交金额为基准, 用复权均价计算成交量
        Volume = Amount / ap;
        AdjustmentCount += number;
    }

    std::vector<MinuteKLine> read_minute_kline_from_csv(const std::string& filename) {
        std::vector<MinuteKLine> klines;
        try {
            // 创建 CSV 读取器
            io::CSVReader<11> in(filename);  // 有 11 列数据

            // 设置表头字段名(用于自动匹配顺序)
            in.read_header(io::ignore_extra_column,
                           "Date", "Open", "Close", "High", "Low",
                           "Volume", "Amount", "Up", "Down", "Datetime", "AdjustmentCount");

            MinuteKLine row = {};
            while (in.read_row(row.Date, row.Open, row.Close, row.High, row.Low, row.Volume, row.Amount,
                               row.Up, row.Down, row.Datetime, row.AdjustmentCount)) {
                klines.emplace_back(row);
            }
        } catch(...) {
            // 忽略异常, 读csv文件失败, 返回空
        }
        return klines;
    }

    std::vector<MinuteKLine> load_minute_kline(const std::string &code, const std::string &freq) {
        auto [minutes, frequency] = pandas::parse_frequency(freq);
        auto filename = config::get_kline_filename_ex(code, frequency);
        spdlog::debug("[dataset::MinuteKLine] kline file: {}", filename);
        return read_minute_kline_from_csv(filename);
    }

    void DataMinuteKLine::Print(const std::string &code, const std::vector<exchange::timestamp> &dates)  {
        (void)code;
        (void)dates;
    }

    void DataMinuteKLine::Update(const std::string &code, const exchange::timestamp &date) {
        if (!mkc_.enabled) {
            return;
        }
//        if(date != exchange::last_trading_day()) {
//            return;
//        }
        const std::string freq_ = mkc_.frequency;
        (void)date;
        // 1. 确定本地有效数据最后1条数据作为拉取数据的开始日期
        auto current_start_date = market_first_date;
        try {
            std::string cache_filename = config::get_kline_filename_ex(code, freq_);
            std::vector<MinuteKLine> cacheMinuteKLines = read_minute_kline_from_csv(cache_filename);
            auto klines_length = cacheMinuteKLines.size();
            int adjust_times = 0; // 除权除息的次数
            auto period = 1;
            auto numberOfDay = 1;
            level1::KLineType kline_type = level1::_1MIN;
            if (mkc_.enabled) {
                period = mkc_.minutes;
                numberOfDay = detail::CN_DEFAULT_TOTALFZNUM / period;
                switch (period) {
                    case 5:
                        kline_type = level1::_5MIN;
                        break;
                    case 15:
                        kline_type = level1::_15MIN;
                        break;
                    case 30:
                        kline_type = level1::_30MIN;
                        break;
                    case 60:
                        kline_type = level1::_1HOUR;
                        break;
                    default:
                        kline_type = level1::_1MIN;
                        break;
                }
            }
            auto klines_offset = detail::MAX_KLINE_LOOKBACK_DAYS * numberOfDay;
            if(klines_length > 0) {
                if (klines_offset > klines_length) {
                    klines_offset = klines_length;
                }
                // 根据最大可以偏移的K线天数, 从缓存中截取对应的日期, 作为从服务器获取数据的起始日期
                const auto& kline = cacheMinuteKLines[klines_length-klines_offset];
                current_start_date = kline.Date; // 修正本次更新的开始日期
                adjust_times = kline.AdjustmentCount;
            }
            // 2. 确定结束日期
            auto current_trading_date = exchange::timestamp::now().pre_market_time();
            spdlog::debug("[dataset::MinuteKLine] [{}]: from {} to {}", code, current_start_date.only_date(), current_trading_date.only_date());
            auto ts = exchange::date_range(current_start_date, current_trading_date);
            auto max_ = 65535;
            auto total = int(ts.size());
            auto max_days = max_/numberOfDay;
            auto days_ = std::min(max_days, total);
            total = days_ * numberOfDay;
            current_start_date = ts[0];
            auto current_end_date = ts[days_-1];
            spdlog::debug("[dataset::MinuteKLine] [{}]: from {} to {}", code, current_start_date.only_date(), current_end_date.only_date());
            u16 step = level1::security_bars_max;
            u16 start = 0;
            // 3. 拉取数据
            std::vector<std::vector<level1::SecurityBar>> hs;
            size_t elementCount = 0;
            do {
                u16 count = step;
                if(total - start >= step) {
                    count = step;
                } else {
                    count = u16(total - start);
                }
                auto reply = detail::fetch_kline(code, start, count, kline_type);
                if (reply.empty()) {
                    break;
                }
                elementCount += reply.size();
                hs.emplace_back(reply);
                if (reply.size() < count) {
                    break;
                }
                start += count;
            } while (start < total);
            // 4. 由于K线数据，每次获取数据是从后往前获取, 所以这里需要反转历史数据的切片
            std::reverse(hs.begin(), hs.end());
            // 5. 调整成交量, 单位从手改成股, vol字段 * 100
            std::vector<MinuteKLine> incremental_klines;
            incremental_klines.reserve(elementCount);
            for(const auto & vec : hs) {
                for (const auto & row : vec) {
                    auto dateTime = exchange::timestamp(row.Year, row.Month, row.Day).pre_market_time();
                    if (dateTime < current_start_date || dateTime > current_trading_date) {
                        // 不在本地更新范围内的记录, 忽略掉
                        continue;
                    }
                    auto kx = MinuteKLine{
                        .Date = dateTime.only_date(), // 日期
                        .Open = row.Open,             // 开盘价
                        .Close = row.Close,           // 收盘价
                        .High = row.High,             // 最高价
                        .Low = row.Low,               // 最低价
                        .Volume = row.Vol * 100,      // 成交量(股)
                        .Amount = row.Amount,         // 成交金额(元)
                        .Up = row.UpCount,            // 上涨家数 / 外盘
                        .Down = row.DownCount,        // 下跌家数 / 内盘
                        .Datetime = row.DateTime,     // 时间
                        .AdjustmentCount = 0      // 新增：除权除息次数
                    };
                    incremental_klines.emplace_back(kx);
                }
            }
            // 6. K线数据转换成MinuteKLine结构
            // 6.1 判断是否已除权的依据
            // 6.1.1 当前更新K线只有1条记录, 则是当前日期, 那么本次更新为当前日期内的多次更新, 需要判断这条新数据是否需要更新缓存以及是否复权
            // 6.1.2 如果隔日更新, 会有2条数据, 缓存中因为偏移是有一条从服务器获取的未复权数据, 第二条数据是当日不需要前复权的记录
            // 6.1.3 只需要判断缓存中的最后一条数据是否除权, 即增量的日线数据的第一条是否需要除权, 如果已除权, 说明缓存内的数据已经全部复权,
            //       只需要复权增量数据的复权即可, 如果没有除权, 则需要对全部的K线数据进行全量处理是否复权
            bool isFreshFetchRequireAdjustment = /*incremental_klines.size() == 1 && */ adjust_times == 1;
            auto dividends = load_xdxr(code);
            if (isFreshFetchRequireAdjustment) {
                // 只除权除息最新的一条记录
                calculate_pre_adjust(incremental_klines, current_start_date, dividends);
            }
            // 6.2 只前复权当日数据
            // 7. 拼接缓存和新增的数据
            std::vector<MinuteKLine> klines;
            // 7.1 先截取本地缓存的数据
            if (klines_length > klines_offset) {
                klines.insert(klines.end(), cacheMinuteKLines.begin(), cacheMinuteKLines.begin()+(klines_length-klines_offset));
            }
            // 7.2 拼接新增的数据
            if (klines.empty()) {
                klines = incremental_klines;
            } else {
                klines.insert(klines.end(), incremental_klines.begin(), incremental_klines.end());
            }
            // 8. 前复权
            if (!isFreshFetchRequireAdjustment) {
                calculate_pre_adjust(klines, current_start_date, dividends);
            }
            // 9. 刷新缓存文件
            save_kline(cache_filename, klines);
        } catch (const std::exception &e) {  // 其他标准异常
            spdlog::error("[dataset::MinuteKLine] - 标准异常: {} (type: {})", e.what(), typeid(e).name());
            // 对于system_error可以记录更多信息
            if (auto se = dynamic_cast<const std::system_error *>(&e)) {
                spdlog::error("[dataset::MinuteKLine] Error code: {}, category: {}", se->code().value(), se->code().category().name());
            }
        } catch (...) {
            spdlog::error("[dataset::MinuteKLine] 获取分钟级别K线异常");
        }
    }

} // namespace datasets

File: quant1x/datasets/kline_minute.rs
use crate::cache::{self, DataAdapter, Kind};
use crate::Timestamp;
use serde::{Deserialize, Serialize};
use std::sync::Arc;

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct MinuteKLine {
    #[serde(rename = "Date")]
    pub date: String,
    #[serde(rename = "Open")]
    pub open: f64,
    #[serde(rename = "Close")]
    pub close: f64,
    #[serde(rename = "High")]
    pub high: f64,
    #[serde(rename = "Low")]
    pub low: f64,
    #[serde(rename = "Volume")]
    pub volume: f64,
    #[serde(rename = "Amount")]
    pub amount: f64,
    #[serde(rename = "Up")]
    pub up: i32,
    #[serde(rename = "Down")]
    pub down: i32,
    #[serde(rename = "Datetime")]
    pub datetime: String,
    #[serde(rename = "AdjustmentCount")]
    pub adjustment_count: i32,
}

impl MinuteKLine {
    pub fn headers() -> Vec<String> {
        vec![
            "Date".into(),
            "Open".into(),
            "Close".into(),
            "High".into(),
            "Low".into(),
            "Volume".into(),
            "Amount".into(),
            "Up".into(),
            "Down".into(),
            "Datetime".into(),
            "AdjustmentCount".into(),
        ]
    }
}

// 推断成交量单位 (参照 C++ 实现)。遍历获取的 SecurityBar 列表, 使用第一条有效记录的
// (Amount / Vol) 与典型价格(平均 OCHL)和 High 比较, 向上取整到 10 的次幂以推断单位。
fn infer_bar_vol_unit(hs: &Vec<Vec<crate::level1::SecurityBar>>) -> f64 {
    for vec in hs.iter() {
        for row in vec.iter() {
            if row.amount <= 0.0 || row.vol <= 0.0 {
                continue;
            }
            let mut typical = (row.open + row.close + row.high + row.low) / 4.0;
            if typical <= 0.0 {
                typical = row.close;
            }
            if typical <= 0.0 {
                continue;
            }
            let implied = row.amount / row.vol;
            if !implied.is_finite() || implied <= 0.0 || row.high <= 0.0 {
                return 1.0;
            }
            if implied <= row.high {
                return 1.0;
            }
            let ratio = implied / row.high;
            if !ratio.is_finite() || ratio <= 1.0 {
                return 1.0;
            }
            let expd = ratio.log10().ceil();
            let mut expi = expd as i32;
            if expi < 0 {
                expi = 0;
            }
            if expi > 9 {
                expi = 9;
            }
            return 10f64.powi(expi);
        }
    }
    1.0
}

#[derive(Debug)]
pub struct DataMinuteKLine;

impl cache::Schema for DataMinuteKLine {
    fn kind(&self) -> Kind {
        crate::datasets::BaseMinuteKLine
    }
    fn owner(&self) -> String {
        crate::cache::DEFAULT_DATA_PROVIDER.to_string()
    }
    fn key(&self) -> String {
        "min".to_string()
    }
    fn name(&self) -> String {
        "分钟K线".to_string()
    }
    fn usage(&self) -> String {
        "分钟K线".to_string()
    }
}

impl DataAdapter for DataMinuteKLine {
    fn print(&self, _code: &str, _dates: &[Timestamp]) {}

    fn update(&self, code: &str, _date: Timestamp) {
        // 读取分钟 K 线配置（必须与 C++ 中的 datasets::get_minute_kline_config 保持一致）
        let mkc = crate::config::get_minute_kline_config();
        if !mkc.enabled {
            log::debug!("[DataMinuteKLine] minute kline not enabled in config");
            return;
        }
        // 使用配置中的频率构建分钟 K 线缓存文件名
        let filename = crate::config::get_kline_filename_ex(code, &mkc.frequency);
        if filename.is_empty() {
            log::error!(
                "[DataMinuteKLine] cannot build minute filename for {}",
                code
            );
            return;
        }
        log::debug!("[DataMinuteKLine] cache filename: {}", filename);

        // 确保父目录存在
        let path = std::path::Path::new(&filename);
        if let Some(parent) = path.parent() {
            if let Err(e) = std::fs::create_dir_all(parent) {
                log::error!(
                    "[DataMinuteKLine] failed to create parent dir {:?}: {}",
                    parent,
                    e
                );
                return;
            }
        }

        // 常量设置（与 C++ 保持一致）
        const MAX_KLINE_LOOKBACK_DAYS: usize = 1;
        const SECURITY_BARS_MAX: usize = 800;
        const CN_DEFAULT_TOTALFZNUM: usize = 240; // default trading minutes in a day

        // 加载本地缓存（如果存在）
        let cache_filename = filename.clone();
        let cache_klines: Vec<MinuteKLine> = read_minute_kline_from_csv(&cache_filename);
        let klines_length = cache_klines.len();
        // 从配置推断周期和对应的 K 线类型
        let period = if mkc.minutes > 0 { mkc.minutes } else { 1 };
        let mut number_of_day = CN_DEFAULT_TOTALFZNUM / period;
        if number_of_day == 0 {
            number_of_day = 1;
        }
        // map period -> level1 category (mirror C++ switch)
        let kline_type: crate::level1::KLineType = match period {
            5 => crate::level1::KLineType::_5Min,
            15 => crate::level1::KLineType::_15Min,
            30 => crate::level1::KLineType::_30Min,
            60 => crate::level1::KLineType::_1Hour,
            _ => crate::level1::KLineType::_1Min, // default
        };

        // Align klines offset to a fixed block size (floor alignment), mirroring the C++ logic.
        // Ensure (klines_length - klines_offset) is an integer multiple of `min_fixed_offset`.
        let min_fixed_offset = MAX_KLINE_LOOKBACK_DAYS * number_of_day;
        let mut klines_offset = min_fixed_offset;
        let mut adjust_times = 0i32;
        // 如果没有缓存，则使用一个非常早的默认日期
        let mut current_start_date =
            crate::Timestamp::pre_market_time(1990, 12, 19).unwrap_or(crate::Timestamp::zero());
        if klines_length > 0 {
            if klines_offset > klines_length {
                klines_offset = klines_length;
            }
            // candidate: 原始候选起点索引
            let candidate = if klines_length > klines_offset {
                klines_length - klines_offset
            } else {
                0
            };
            // 使用 floor 对齐到 min_fixed_offset 的倍数，确保 (klines_length - klines_offset) 为该块大小的整数倍
            let mut aligned = if min_fixed_offset > 0 {
                (candidate / min_fixed_offset) * min_fixed_offset
            } else {
                0
            };
            // 边界保护
            if aligned >= klines_length {
                aligned = 0;
            }
            // 重新计算 klines_offset，使得 klines_length - klines_offset == aligned
            klines_offset = klines_length - aligned;
            // 根据对齐后的索引取出对应的日期作为拉取起点，并保留该边界行的 adjustment_count
            let kline = &cache_klines[aligned];
            // kline.date 是字符串, 尝试解析并转换为盘前时间
            if let Ok(mut ts) = crate::Timestamp::parse(&kline.date) {
                ts = crate::Timestamp::pre_market_time_from_current(&ts).unwrap_or(ts);
                current_start_date = ts;
            }
            adjust_times = kline.adjustment_count;

            // 如果 aligned 看起来不是某个交易日的首条记录，记录警告以便人工审查。
            if aligned > 0 && cache_klines[aligned - 1].date == cache_klines[aligned].date {
                log::warn!(
                    "[DataMinuteKLine] aligned index {} is not day-first for {} (date={})",
                    aligned,
                    code,
                    cache_klines[aligned].date
                );
            }
        }

        // 构建从起始日期到今日盘前的日期范围
        let mut current_end_date =
            crate::Timestamp::pre_market_time_from_current(&crate::Timestamp::now())
                .unwrap_or(crate::Timestamp::now());
        let ts_range = crate::exchange::date_range(current_start_date, current_end_date, false);
        if ts_range.is_empty() {
            log::debug!("[DataMinuteKLine] empty date range for {}", code);
            return;
        }
        log::info!(
            "[DataMinuteKLine] updating {} from {} to {} ({} days) with period {} minutes",
            code,
            current_start_date.only_date(),
            current_end_date.only_date(),
            ts_range.len(),
            period
        );
        // 与 C++ 行为对齐：将分钟总条目数限制在 u16 最大值（65535），并使用 number_of_day 将天数转换为分钟条目数
        let max_entries: usize = 65535;
        let total_days = ts_range.len();
        let max_days = if number_of_day > 0 {
            max_entries / number_of_day
        } else {
            total_days
        };
        let days = std::cmp::min(max_days, total_days);
        if days == 0 {
            log::debug!("[DataMinuteKLine] empty date range for {}", code);
            return;
        }
        let total = days * number_of_day;
        // 从后往前取days的交易日期
        current_start_date = ts_range[total_days - days];
        current_end_date = ts_range[total_days - 1];
        log::info!(
            "[DataMinuteKLine] fetching {} days from {} to {} ({} entries)",
            days,
            current_start_date.only_date(),
            current_end_date.only_date(),
            total
        );
        // 从 level1 分页拉取分钟数据（C++ 中日线用类目 9；分钟线依据频率使用 1..8，这里简化使用 1 作为分钟类目）
        let mut hs: Vec<Vec<crate::level1::SecurityBar>> = Vec::new();
        let step = SECURITY_BARS_MAX;
        let mut start_idx: usize = 0;
        while start_idx < total {
            let remaining = total - start_idx;
            let count = std::cmp::min(step, remaining) as u16;
            // 最低可观测性：在拉取每页数据前记录请求参数
            log::info!(
                "[DataMinuteKLine] fetch request: code={} kline_type={:?} start_idx={} count={} total={}",
                code,
                kline_type,
                start_idx,
                count,
                total
            );
            match crate::datasets::kline_raw::fetch_kline(code, start_idx as u32, count, kline_type)
            {
                Some(resp) if !resp.list.is_empty() => {
                    let response_len = resp.list.len();
                    hs.push(resp.list);
                    if response_len < count as usize {
                        break;
                    }
                    start_idx = start_idx.saturating_add(count as usize);
                }
                _ => {
                    log::warn!(
                        "[DataMinuteKLine] fetch_kline returned empty for {} start={}",
                        code,
                        start_idx
                    );
                    break;
                }
            }
        }

        if hs.is_empty() {
            // 空列表不保存CSV，包括表头
            return;
        }

        // reverse pages to ascending time (C++ behavior)
        hs.reverse();

        let mut incremental_klines: Vec<MinuteKLine> = Vec::new();
        let bar_vol_unit = infer_bar_vol_unit(&hs);
        for page in hs.iter() {
            for row in page.iter() {
                let date_time =
                    crate::Timestamp::pre_market_time(row.year, row.month as u32, row.day as u32)
                        .unwrap_or(crate::Timestamp::now());
                if date_time < current_start_date || date_time > current_end_date {
                    continue;
                }
                let kx = MinuteKLine {
                    date: date_time.only_date(),
                    open: row.open,
                    close: row.close,
                    high: row.high,
                    low: row.low,
                    volume: row.vol * bar_vol_unit,
                    amount: row.amount,
                    up: row.up_count as i32,
                    down: row.down_count as i32,
                    datetime: row.datetime.clone(),
                    adjustment_count: 0,
                };
                incremental_klines.push(kx);
            }
        }

        let is_fresh_fetch_require_adjustment = adjust_times == 1;
        let dividends = crate::datasets::xdxr::load_xdxr(code);
        if is_fresh_fetch_require_adjustment {
            apply_forward_adjustment_for_event!(&mut incremental_klines, current_start_date, &dividends);
        }

        // merge cache and incremental
        let mut klines: Vec<MinuteKLine> = Vec::new();
        if klines_length > klines_offset {
            klines.extend_from_slice(&cache_klines[..(klines_length - klines_offset)]);
        }
        if klines.is_empty() {
            klines = incremental_klines.clone();
        } else {
            klines.extend(incremental_klines.into_iter());
        }

        if !is_fresh_fetch_require_adjustment {
            apply_forward_adjustment_for_event!(&mut klines, current_start_date, &dividends);
        }

        // persist
        if klines.is_empty() {
            // 空列表不保存CSV，包括表头
            return;
        }
        let tmp = format!("{}.tmp", filename);
        match std::fs::File::create(&tmp) {
            Ok(f) => {
                let mut w = csv::Writer::from_writer(f);
                if let Err(e) = w.write_record(MinuteKLine::headers()) {
                    log::error!("[DataMinuteKLine] write header failed: {}", e);
                }
                for row in klines.iter() {
                    let rec: Vec<String> = vec![
                        row.date.clone(),
                        row.open.to_string(),
                        row.close.to_string(),
                        row.high.to_string(),
                        row.low.to_string(),
                        row.volume.to_string(),
                        row.amount.to_string(),
                        row.up.to_string(),
                        row.down.to_string(),
                        row.datetime.clone(),
                        row.adjustment_count.to_string(),
                    ];
                    if let Err(e) = w.write_record(rec) {
                        log::error!("[DataMinuteKLine] write row failed: {}", e);
                    }
                }
                let _ = w.flush();
                if let Err(e) = std::fs::rename(&tmp, &filename) {
                    log::error!(
                        "[DataMinuteKLine] rename failed {} -> {}: {}",
                        tmp,
                        filename,
                        e
                    );
                }
            }
            Err(e) => {
                log::error!("[DataMinuteKLine] create tmp {} failed: {}", tmp, e);
            }
        }
    }
}

// read minute CSV
fn read_minute_kline_from_csv(filename: &str) -> Vec<MinuteKLine> {
    let mut klines: Vec<MinuteKLine> = Vec::new();
    match std::fs::File::open(filename) {
        Ok(f) => {
            let mut rdr = csv::ReaderBuilder::new().has_headers(true).from_reader(f);
            match rdr
                .deserialize::<MinuteKLine>()
                .collect::<Result<Vec<MinuteKLine>, csv::Error>>()
            {
                Ok(v) => klines = v,
                Err(e) => log::error!(
                    "[DataMinuteKLine] failed to deserialize minute file {}: {}",
                    filename,
                    e
                ),
            }
        }
        Err(_) => {}
    }
    klines
}

pub fn init() {
    let plugin = Arc::new(DataMinuteKLine) as Arc<dyn DataAdapter>;
    cache::register(plugin);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_minute_kline_update() {
        let adapter: DataMinuteKLine = DataMinuteKLine;
        let code = "sh510050";
        let date = Timestamp::now();
        adapter.update(code, date);
    }
}
